tests/test_cache.py

```diff
@@ -0,0 +1,105 @@
+import httpretty
+import redis
+import shutil
+import unittest
+import requests
+from cachecontrol.caches.file_cache import FileCache
+from cachecontrol.caches.redis_cache import RedisCache
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+
+file_cache_directory = ".testcache"
+
+
+def create_redis_connection_pool(
+    redis_port="6379", redis_host="localhost", redis_db=0
+):
+    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)
+
+
+class TestCachedSession(unittest.TestCase):
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
+        except:
+            pass
+
+    @httpretty.activate
+    def test_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(
+            fallback_cache_duration=10,
+            file_cache_directory=file_cache_directory,
+        )
+        resp = session.get("https://httpbin.org/ip")
+
+        for h in resp.headers:
+            if h == "cache-control":
+                print(resp.headers[h])
+
+        self.assertIn(
+            "Automatically cached for 0:00:10", resp.headers["warning"]
```

Comment:
Here you're definitely testing internals, and not testing externals.

The "warning" header and its contents are completely custom items that happen to be added by CacheControl. They are of no functional use to the client, they don't prove that the response will actually be cached as desired, and testing them means that if ever anyone swaps out CacheControl for a different module, or possibly if ever CacheControl is upgraded, the test will fail.

What I envisaged here was to make multiple `session.get` calls to the same URL, and check the response was returned from the cache or not, as expected. You could use the `from_cache` property, but even that is a little bit testing internals. Even better would be to update the content returned by the endpoint. Maybe we could write a custom `body=` callback function that returns the epoch time? We could then easily prove that the request was a cached version, or not.

Reply:
Oooh, http://now.httpbin.org/ is a good endpoint for these tests, which we could then mock similarly to how it works live =)